File: win-prog/ProgWinCS/Brushes and Pens/HexagonGradientBrush/HexagonGradientBrush.cs
//---------------------------------------------------
// HexagonGradientBrush.cs © 2001 by Charles Petzold
//---------------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

class HexagonGradientBrush: PrintableForm
{
     const float fSide = 50;       // Side (also radius) of hexagon

     public new static void Main()
     {
          Application.Run(new HexagonGradientBrush());
     }
     public HexagonGradientBrush()
     {
          Text = "Hexagon Gradient Brush";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
               // Calculate half the hexagon height.

          float fHalf = fSide * (float) Math.Sin(Math.PI / 3); 
          
               // Define a hexagon including some extra width

          PointF[] aptf = {new PointF( fSide,         0),           
                           new PointF( fSide * 1.5f,  0),
                           new PointF( fSide,         0),           
                           new PointF( fSide / 2,    -fHalf),
                           new PointF(-fSide / 2,    -fHalf),
                           new PointF(-fSide,         0),
                           new PointF(-fSide * 1.5f,  0), 
                           new PointF(-fSide,         0),
                           new PointF(-fSide / 2,     fHalf),     
                           new PointF( fSide / 2,     fHalf) };

               // Create the first brush.

          PathGradientBrush pgbrush1 =
                              new PathGradientBrush(aptf, WrapMode.Tile);

               // Offset the hexagon and define the second brush.

          for (int i = 0; i < aptf.Length; i++)
          {
               aptf[i].X += fSide * 1.5f;
               aptf[i].Y += fHalf;
          }
          PathGradientBrush pgbrush2 = 
                              new PathGradientBrush(aptf, WrapMode.Tile);

          grfx.FillRectangle(pgbrush1, 0, 0, cx, cy);
          grfx.FillRectangle(pgbrush2, 0, 0, cx, cy);
     }
}

File: win-prog/ProgWinVB/Brushes and Pens/HexagonGradientBrush/HexagonGradientBrush.vb
'-----------------------------------------------------
' HexagonGradientBrush.vb (c) 2002 by Charles Petzold
'-----------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Class HexagonGradientBrush
    Inherits PrintableForm

    Const fSide As Single = 50    ' Side (also radius) of hexagon

    Shared Shadows Sub Main()
        Application.Run(New HexagonGradientBrush())
    End Sub

    Sub New()
        Text = "Hexagon Gradient Brush"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        ' Calculate half the hexagon height.
        Dim fHalf As Single = CSng(fSide * Math.Sin(Math.PI / 3))

        ' Define a hexagon including some extra width.
        Dim aptf() As PointF = {New PointF(fSide, 0), _
                                New PointF(fSide * 1.5, 0), _
                                New PointF(fSide, 0), _
                                New PointF(fSide / 2, -fHalf), _
                                New PointF(-fSide / 2, -fHalf), _
                                New PointF(-fSide, 0), _
                                New PointF(-fSide * 1.5, 0), _
                                New PointF(-fSide, 0), _
                                New PointF(-fSide / 2, fHalf), _
                                New PointF(fSide / 2, fHalf)}

        ' Create the first brush.
        Dim pgbr1 As PathGradientBrush = _
                            New PathGradientBrush(aptf, WrapMode.Tile)

        ' Offset the hexagon and define the second brush.
        Dim i As Integer
        For i = 0 To aptf.GetUpperBound(0)
            aptf(i).X += fSide * 1.5F
            aptf(i).Y += fHalf
        Next i
        Dim pgbr2 As PathGradientBrush = _
                            New PathGradientBrush(aptf, WrapMode.Tile)

        grfx.FillRectangle(pgbr1, 0, 0, cx, cy)
        grfx.FillRectangle(pgbr2, 0, 0, cx, cy)
    End Sub
End Class
